src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -651,9 +651,15 @@ private static bool LogFailure(object hostNameOrAddress, long? startingTimestamp
         /// than having all concurrent requests for the same host share the exact same task, so that any shuffling of the results
         /// by the OS to enable round robin is still perceived.
         /// </remarks>
-        private static Task<TResult> RunAsync<TResult>(Func<object, long, TResult> func, object key, CancellationToken cancellationToken)
+        private static Task<TResult> RunAsync<TResult>(Func<object, NameResolutionActivity, TResult> func, object key, CancellationToken cancellationToken)
         {
-            long startingTimestamp = NameResolutionTelemetry.Log.BeforeResolution(key);
+            Activity? activityToRestore = NameResolutionActivity.IsTracingEnabled() ? Activity.Current : null;
+            NameResolutionActivity activity = NameResolutionTelemetry.Log.BeforeResolution(key);
+            if (NameResolutionActivity.IsTracingEnabled())
+            {
+                // Do not overwrite Activity.Current in the caller's ExecutionContext.
```

Comment:
?

Reply:
> so we cannot disable tracing

I'm not following. The IsTracingEnabled method is doing:
```
s_activitySource.HasListeners()
```
What prevents the last listener from unregistering before this if block?

Reply:
I misunderstood you and thought you want me to avoid the manual reset of `Activity.Current`.

> What prevents the last listener from unregistering before this if block?

Yes, that's a problem, will fix it.

Reply:
Usually trace collectors are interested in the Activity `Stop` event. So, if the listeners stopped listening before Activity `Stop` event, the whole tracing in the `Activity.Current` will be lost as no-one interested in collecting. 

> Unfortunately there is no way to start a new activity without overwriting Activity.Current.
> shouldn't this be a supported scenario?

I am not sure I understand the scenario here, if you start a new `Activity`, the newly started activity will be set to `Activity.Current`. When you stop it, the original  `Activity.Current` will be restored again. What exactly do you want to achieve here?

CC @noahfalk

Reply:
@tarekgh the value of `Activity.Current` depends on the `ExecutionContext` since it's being stored in an `AsyncLocal`. Here we are starting the activity in the parent context (sync path) before kicking off the child Task where the activity is being stopped. This means Start and Stop will manipulate different AsyncLocals.

Reply:
Thanks. 

I was wondering about your statement `Unfortunately there is no way to start a new activity without overwriting Activity.Current.`. So, I was wondering what is your scenario that needs to start activity and not setting it to the Activity.Current?

Reply:
Maybe it's more clear with code:

```C#
public static Task MyAsyncApi()
{
    Activity? activity = new Activity("MyAsyncApi").Start();

    DoSomeSyncStuff();

    return Task.Run(Impl);

    void Impl()
    {
        DoBlockingIO();
        activity?.Stop();
    }
}


Task task = MyAsyncApi();
Console.WriteLine(Activity.Current.OperationName); // prints MyAsyncApi

await task;
Console.WriteLine(Activity.Current.OperationName); // still prints MyAsyncApi :(((
```

The only way to ensure that `Activity.Current` will not be set to `MyAsyncApi` on the parent context forever is to reset `Activity.Current` after starting the `MyAsyncApi` Activity.

Reply:
I'm not sure what improvements we could realisticaly make? Its an invariant of the AsyncLocal design that the value doesn't change across an async call. I agree with you that it feels awkward/asymmetric that thread-local values flow out of synchronous calls but async-local values do not flow out of async calls, but its hard for me to imagine that design would change in the future.

For now I think the guidance is:
- Preferably start and stop your activities within the same frame. I'd guess 95+% of usage lands here.
- Failing that, be aware of the restrictions in AsyncLocal flow and account for them as needed